src/finalizer/finalizer.cpp

```diff
@@ -5,207 +5,325 @@
+                        if (WIU_RESTART_INITIATED == restart || WIU_RESTART_REQUIRED == restart)
+                        {
+                            LogStringLine(REPORT_STANDARD, "  Reboot requested, deferring.");
+                            *pbRestartRequired = TRUE;
+                        }
+
+                        // Reset potential failures so we can continue to remove as many dependents as possible.
+                        hr = S_OK;
+                    }
+                    else if (HRESULT_FROM_WIN32(ERROR_UNKNOWN_PRODUCT) == hr || HRESULT_FROM_WIN32(ERROR_UNKNOWN_PROPERTY) == hr)
+                    {
                         // Possibly a corrupted provider key that wasn't cleaned up. We'll just ignore it.
-						LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
-						hr = S_OK;
-					}
-				}
-			}
-		}
+                        LogStringLine(REPORT_STANDARD, "  Product is not installed, ProductCode:%ls, result: 0x%.8x", sczProductId, hr);
+                        hr = S_OK;
+                    }
+                }
+            }
+        }
 
-		ReleaseRegKey(hkDependentsKey);
-		ReleaseRegKey(hkProviderKey);
-	}
+        ReleaseRegKey(hkDependentsKey);
+        ReleaseRegKey(hkProviderKey);
+    }
 
 LExit:
-	ReleaseStr(sczProductName);
-	ReleaseStr(sczProductId);
-	ReleaseStr(sczProviderKey);
-	ReleaseStr(sczDependentsKey);
-	ReleaseRegKey(hkDependentsKey);
-	ReleaseRegKey(hkProviderKey);
-	ReleaseRegKey(hkInstallerDependenciesKey);
-	return hr;
+    ReleaseStr(sczProductName);
+    ReleaseStr(sczProductId);
+    ReleaseStr(sczProviderKey);
+    ReleaseStr(sczDependentsKey);
+    ReleaseRegKey(hkDependentsKey);
+    ReleaseRegKey(hkProviderKey);
+    ReleaseRegKey(hkInstallerDependenciesKey);
+    return hr;
+}
+
+extern "C" HRESULT ParseSdkVersion(LPWSTR sczSdkVersion, INT * piMajor, INT * piMinor, INT * piFeatureBand)
+{
+    HRESULT hr = S_OK;
+    UINT cVersionParts = 0;
+    LPWSTR* rgsczVersionParts = NULL;
+
+    hr = StrSplitAllocArray(&rgsczVersionParts, &cVersionParts, sczSdkVersion, L".");
+    ExitOnFailure(hr, "Failed to split version.");
+
+    // We only care about the major.minor.patch values
+    // to convert to a feature band. If we don't have at least
+    // all 3 parts, we'll ignore the value.
+    if (3 > cVersionParts)
+    {
+        ExitOnFailure(E_INVALIDARG, "Invalid SDK version: %ls %li", sczSdkVersion, cVersionParts);
+    }
+
+    int iMajor = 0;
+    int iMinor = 0;
+    int iPatch = 0;
+
+    hr = StrStringToInt32(rgsczVersionParts[0], 0, piMajor);
+    ExitOnFailure(hr, "Invalid major version.");
+    hr = StrStringToInt32(rgsczVersionParts[1], 0, piMinor);
+    ExitOnFailure(hr, "Invalid minor version.");
+
+    // If this is a valid SDK version the 'patch' should be a 3 digit field
+    // containing the feature band and patch level, e.g. 100 or 207. We 
+    // can discard any prerelease labels from the semantic version.
```

Comment:
What happens if the third version part does include semantic prerelease labels (or build version information).  Does `StrStringToInt32` fail if it encounters a `-` or `+`, or does it just parse the string up to there and return it?

EDIT: OK, I see that you're passing 3 as the second argument.  Does that mean just parse three characters, and 0 means the whole string?

Reply:
Correct, this is just existing libraries from the WiX SDK I'm using. We expect that the SDK patch would contain 3 digits, if not, it's not a valid featureband+patch combination, e.g. if someone defined a version like 5.3.62.